src/ray/raylet/node_manager.cc

```diff
@@ -1442,13 +1399,72 @@ void NodeManager::HandleSubmitTask(const rpc::SubmitTaskRequest &request,
                                    rpc::SendReplyCallback send_reply_callback) {
   rpc::Task task;
   task.mutable_task_spec()->CopyFrom(request.task_spec());
+  // Set the caller's node ID.
+  if (task.task_spec().caller_address().raylet_id() == "") {
+    task.mutable_task_spec()->mutable_caller_address()->set_raylet_id(
+        gcs_client_->client_table().GetLocalClientId().Binary());
+  }
 
   // Submit the task to the raylet. Since the task was submitted
   // locally, there is no uncommitted lineage.
   SubmitTask(Task(task), Lineage());
   send_reply_callback(Status::OK(), nullptr, nullptr);
 }
 
+void NodeManager::HandleWorkerLeaseRequest(const rpc::WorkerLeaseRequest &request,
+                                           rpc::WorkerLeaseReply *reply,
+                                           rpc::SendReplyCallback send_reply_callback) {
+  rpc::Task task_message;
+  task_message.mutable_task_spec()->CopyFrom(request.resource_spec());
+
+  // Override the task dispatch to call back to the client instead of executing the
+  // task directly on the worker.
+  Task task(task_message);
+  RAY_LOG(DEBUG) << "Worker lease request " << task.GetTaskSpecification().TaskId();
+  TaskID task_id = task.GetTaskSpecification().TaskId();
+  task.OnDispatchInstead(
+      [this, task_id, reply, send_reply_callback](const std::shared_ptr<void> granted,
+                                                  const std::string &address, int port) {
+        RAY_LOG(DEBUG) << "Worker lease request DISPATCH " << task_id;
+        reply->mutable_worker_address()->set_ip_address(address);
+        reply->mutable_worker_address()->set_port(port);
+        reply->mutable_worker_address()->set_raylet_id(
+            gcs_client_->client_table().GetLocalClientId().Binary());
+        send_reply_callback(Status::OK(), nullptr, nullptr);
+
+        // TODO(swang): Release worker if other end hangs up (can be done with
+        // lease timeout).
```

Comment:
We should probably kill any leased workers in this scenario.

Reply:
I split it into 2 TODOs, one for when the other end hangs up and the other for if the lease term expires. I'm not sure if it makes sense to kill the worker if it's only the latter case. Seems like it could happen more often, and we could ask the owner to return the worker instead.